File: lintcode/permutations-ii.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/permutations-ii/

class Solution:
    def helper(self, nums, visited, path, ans):
        if len(path) == len(nums):
            ans.append(path[:])
            return

        existed = set()
        for i in xrange(len(nums)):
            if i in visited or nums[i] in existed:
                continue

            visited.add(i)
            existed.add(nums[i])
            path.append(nums[i])
            self.helper(nums, visited, path, ans)
            path.pop()
            visited.remove(i)

    """
    @param nums: A list of integers.
    @return: A list of unique permutations.
    """
    def permuteUnique(self, nums):
        # write your code here
        if not nums:
            return []

        ans = []
        self.helper(nums, set(), [], ans)
        return ans

if __name__ == '__main__':
    s = Solution()
    print s.permuteUnique([1,2,2])

File: leetcode-2/permutations-ii.cc
// Question: https://leetcode.com/problems/permutations-ii/

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        unordered_set<int> visited;
        vector<int> path;
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());
        helper(nums, visited, path, ans);
        return ans;
    }

    void helper(vector<int>& nums, unordered_set<int> visited, vector<int>& path, vector<vector<int>>& ans) {
        if (path.size() == nums.size()) {
            ans.push_back(path);
        } else if (path.size() < nums.size()) {
            for (int i = 0; i < nums.size(); ++i) {
                if (visited.count(i) == 0) {
                    if (i > 0 && nums[i] == nums[i-1] && visited.count(i-1) == 0) {
                        continue;
                    }
                    visited.insert(i);
                    path.push_back(nums[i]);
                    helper(nums, visited, path, ans);
                    path.pop_back();
                    visited.erase(i);
                }
            }
        }
    }
};
